examples/Nat/TcpState.cs

```diff
@@ -0,0 +1,126 @@
+/*
+Pax : tool support for prototyping packet processors
+Jonny Shipton, Cambridge University Computer Lab, July 2016
+
+Use of this source code is governed by the Apache 2.0 license; see LICENSE.
+*/
+
+using System;
+using PacketDotNet;
+
+namespace Pax.Examples.Nat
+{
+  /// <summary>
+  /// Represents the state of a TCP connection.
+  /// </summary>
+  internal class TcpState : ITransportState<TcpPacket>
+  {
+    /// <summary>
+    /// The time for which a TCP connection should be kept open after the remote host terminates it with a Fin.
+    /// </summary>
+    private readonly TimeSpan TIME_WAIT;
+
+    /// <summary>
+    /// The inferred state of the TCP connection from the inside node to the outside node. Related to the TCP state on the outside node.
+    /// </summary>
+    private TcpDirectionalState InOutConnection = TcpDirectionalState.None;
+
+    /// <summary>
+    /// The inferred state of the TCP connection from the outside node to the inside node. Related to the TCP state on the inside node.
+    /// </summary>
+    private TcpDirectionalState OutInConnection = TcpDirectionalState.None;
+
+    /// <summary>
+    /// The time that the connection was closed, if applicable.
+    /// </summary>
+    private DateTime? CloseTime = null;
+
+    /// <summary>
+    /// True if the TCP connection from the inside to the outside is closed.
+    /// </summary>
+    public bool ClosedFromInside { get { return InOutConnection == TcpDirectionalState.None || InOutConnection == TcpDirectionalState.FinAck; } }
+
+    /// <summary>
+    /// True if the TCP connection from the outside to the inside is closed.
+    /// </summary>
+    public bool ClosedFromOutside { get { return OutInConnection == TcpDirectionalState.None || OutInConnection == TcpDirectionalState.FinAck; } }
+
+    /// <summary>
+    /// True if the connection is in the TIME_WAIT state.
+    /// </summary>
+    public bool InTimeWait { get { return CloseTime.HasValue && DateTime.Now - CloseTime < TIME_WAIT; } }
+
+    /// <summary>
+    /// True if the TCP connections in both directions are closed, determining if the connection can be removed before the timeout elapses.
+    /// </summary>
+    public bool CanBeClosed { get { return ClosedFromInside && ClosedFromOutside && !InTimeWait; } }
+
+    /// <summary>
+    /// Creates a new TcpState in the initial state.
+    /// </summary>
+    /// <param name="time_wait">The duration of the TIME_WAIT state.</param>
+    public TcpState(TimeSpan time_wait)
+    {
+      TIME_WAIT = time_wait;
+    }
+
+    public void UpdateState(TcpPacket packet, bool packetFromInside)
```

Comment:
This function is quote important, and deserves a couple of sentences of comment describing its intended behaviour. It's not immediately obvious from the code how the state is being transitioned and what role the parameters play.